### PR TITLE
AffineConstraints modernization

### PR DESCRIPTION
**Pull request by drwells**
I compiled the library with the `Wnull-dereference` flag earlier and it found a few places where we are using pointers in AffineConstraints. In most of these places we can make things a bit clearer by just using references.

I am not convinced that `clang-tidy` was ever run on this file: I used it to modernize the loops (I also had to convert a few by hand).

### CODE REVIEW

**Comment by drwells** [comment]
That would explain it.

`modernize-loop-convert` also made some inexplicable type conversions, i.e., it wrote `unsigned int` instead of `size_type`, so perhaps its for the best that we keep this one out.

File: include/deal.II/lac/affine_constraints.templates.h [hunk 1]
```diff
@@ -334,11 +312,11 @@ AffineConstraints<number>::close()
...
     size_type              counter = 0;
-    for (typename std::vector<ConstraintLine>::const_iterator line =
-           lines.begin();
-         line != lines.end();
-         ++line, ++counter)
-      new_lines[calculate_line_index(line->index)] = counter;
+    for (const ConstraintLine &line : lines)
+      {
+        new_lines[calculate_line_index(line.index)] = counter;
+        ++counter;
```
Comment (masterleinad) [inline include/deal.II/lac/affine_constraints.templates.h]: You could shorten this to `new_lines[calculate_line_index(line.index)] = counter++;` for better performance. :wink:

Comment (drwells) [reply include/deal.II/lac/affine_constraints.templates.h]: I have trouble remembering the semantics of ` a = x++` vs `a = ++x` so yes, I would also prefer to keep them separate :)

Comment (masterleinad) [reply include/deal.II/lac/affine_constraints.templates.h]: Sure, that's fine with me.

File: include/deal.II/lac/affine_constraints.templates.h [hunk 2]
```diff
@@ -71,16 +71,6 @@ AffineConstraints<number>::copy_from(const AffineConstraints<number> &other)
 
 
 
-template <typename number>
-bool
-AffineConstraints<number>::check_zero_weight(
-  const std::pair<size_type, number> &p)
-{
-  return (p.second == number(0.));
-}
```
[the remark this answers is not in the document]
Comment (drwells) [reply include/deal.II/lac/affine_constraints.templates.h]: I completely missed that this is a static member function. Fixed.

**Comment by masterleinad** [comment]
Thanks!

**Comment by drwells** [comment]
```
/home/bob/source/bundled/boost-1.62.0/include/boost/serialization/level.hpp:22:10: fatal error: cannot open file '/home/bob/source/bundled/boost-1.62.0/include/boost/type_traits/is_fundamental.hpp': Invalid argument
#include <boost/type_traits/is_fundamental.hpp>
         ^
1 error generated.
ninja: build stopped: subcommand failed.
build FAILED
```
as if we were not having enough other issues with serialization :(

**Comment by bangerth** [comment]
One of the tester got stuck on a missing file, but I think that's unrelated. The file exists in our tree. So I'll merge -- the other tester succeeded with all of its 3661 tests, after all.
